torch/distributed/distributed_c10d.py

```diff
@@ -3603,16 +3621,27 @@ def _new_group_with_tag(ranks=None, timeout=default_pg_timeout, backend=None, pg
         global_rank: group_rank for group_rank, global_rank in enumerate(ranks)
     }
 
-    # barrier at the end to ensure that once we return from this method, all
-    # process groups including global variables are updated correctly on all
-    # ranks.
-    if backend == Backend.MPI:
-        # MPI doesn't have store.
-        barrier()
+    if _barrier_after_init == 1:
+        # barrier at the end to ensure that once we return from this method, all
+        # process groups including global variables are updated correctly on all
+        # ranks.
+        # Update 04/2023: for large-scale runs, these barriers (esp. store-based
+        # barrier) may be costly and/or unscalable. Also, in a lot of cases,
+        # these barriers may be unnecessary, as proved by a green CI after
+        # removal. An environment variable `TORCH_DIST_INIT_BARRIER` has been
+        # added which, when set to 0, will disable these barriers.
+        if backend == Backend.MPI:
+            # MPI doesn't have store.
+            barrier()
+        else:
+            # Use store based barrier here since barrier() used a bunch of
+            # default devices and messes up NCCL internal state.
+            _store_based_barrier(global_rank, default_store, timeout)
```

Comment:
I believe that this is the only place that `_store_based_barrier` is used and its not a public api, so if we are guarding it here with the `_barrier_after_init` flag, we also don't need to make any logic changes in the function right?

Reply:
> we also don't need to make any logic changes in the function right?
If by "the function" you mean `_store_based_barrier`, then I believe so.